src/dstack/_internal/server/background/tasks/process_instances.py

```diff
@@ -88,6 +117,136 @@ async def process_instances() -> None:
         PROCESSING_POOL_IDS.difference_update(i.id for i in instances)
 
 
+async def add_remote(instance_id: UUID) -> None:
+    async with get_session_ctx() as session:
+        instance = (
+            await session.scalars(
+                select(InstanceModel)
+                .where(InstanceModel.id == instance_id)
+                .options(joinedload(InstanceModel.project))
+            )
+        ).one()
+
+        retry_duration_deadline = instance.created_at.replace(
+            tzinfo=datetime.timezone.utc
+        ) + timedelta(seconds=PROVISIONING_TIMEOUT_SECONDS)
+        if retry_duration_deadline < get_current_datetime():
+            instance.status = InstanceStatus.TERMINATED
+            instance.deleted = True
+            instance.deleted_at = get_current_datetime()
+            instance.termination_reason = "The proivisioning timeout expired"
+            await session.commit()
+            logger.warning(
+                "Failed to start the instance in %s seconds. Terminate instance %s",
+                PROVISIONING_TIMEOUT_SECONDS,
+                instance.name,
+                extra={
+                    "instance_name": instance.name,
+                    "instance_status": InstanceStatus.TERMINATED.value,
+                },
+            )
+            return
+
+        try:
+            remote_details = RemoteConnectionInfo.parse_raw(
+                cast(str, instance.remote_connection_info)
+            )
+
+            # Prepare connection key
+            try:
+                private_string = [
+                    sk.private for sk in remote_details.ssh_keys if sk.private is not None
+                ][0]
+            except IndexError:
+                logger.error("There are no ssh private key")
+                raise BackendError()
```

Comment:
BackendError() is there to be raised by backends. Weird to see it raised in process_instances.py.

Reply:
Fixed